### PR TITLE
Add napari.imshow to return viewer and layers

### PR DESCRIPTION
**Pull request by kcpevey**
# Description

We have to main options for returning a viewer or the layers when we're adding an image:
```
viewer = napari.view_image(image)
image_layer = viewer.layers[0]
```
or,
```
viewer = napari.Viewer()
image_layer = viewer.add_image(image)
```

This PR adds a new method, `napari.imshow` which returns _both_ a `viewer` and an `image_layer`

Partially fulfills https://github.com/napari/napari/issues/4693

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update
  - Not sure on this...

# References

https://github.com/napari/napari/issues/4693

# How has this been tested?

- [x] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by kcpevey** [comment]
I'm unsure about the best way to handle kwargs here. I started by splitting off the viewer kwargs like is done in the `view_` methods, but that didn't cover everything. Its a little awkward. 

I'm also unsure about the `multichannel` and `channel_axis`. Its not well documented. I gather from the tests that I can add `channel_axis` and it will infer `multichannel`. If I haven't handled this correctly, please let me know.

**Comment by kcpevey** [comment]
It was mentioned [here](https://github.com/napari/napari/issues/4693#issuecomment-1200509018) that this function should return a list of all the added layers. In order to do that I'd need to make `data` into a list, but `data` can be an array or list of arrays. Which means I'd have no way to distinguish which one is intended as multiple layers and which is multi-dimensional? Right?

**Comment by liu-ziyang** [comment]
> Which means I'd have no way to distinguish which one is intended as multiple layers and which is multi-dimensional? Right?

being naive here, would it be sufficient to isinstance check the first element being Layer or not?

**Comment by kcpevey** [comment]
> would it be sufficient to isinstance check the first element being Layer or not?

this is before the Layer instance is created. I'd be inspecting the `data` variable going into `add_image` which I believe can be an array or list of arrays. 

Suppose my data is a list of arrays, how will the `imshow` method know if I want this to be added as one single layer or as multiple layers?

Its probably worth noting that I'm also naive and I'm not 100% confident I know how all this works! :)

**Comment by tlambert03** [comment]
my first thought is that this ends up being pretty much identical to all the other `_make_viewer_then`... functions in this module, except that 1) you need to be able to pass in an optional viewer and 2) you need to be able to return the added layers.  So, what do you think about:

1. modify `_make_viewer_then` slightly to read as follows:
    ```python
    def _make_viewer_then(add_method: str, args, kwargs, viewer=None) -> Tuple[Viewer, Any]:
        """Utility function that creates a viewer, adds a layer, returns viewer."""
        vkwargs = {k: kwargs.pop(k) for k in list(kwargs) if k in _viewer_params}
        # separate dims kwargs because we want to set those after adding data
        dims_kwargs = {
            k: vkwargs.pop(k) for k in list(vkwargs) if k in _dims_params
        }
        if viewer is None:
            viewer = Viewer(**vkwargs)
        kwargs.update(kwargs.pop("kwargs", {}))
        method = getattr(viewer, add_method)
        added = method(*args, **kwargs)
        for arg_name, arg_val in dims_kwargs.items():
            setattr(viewer.dims, arg_name, arg_val)
        return viewer, added
    ```
   (this changes the signature to accept an optional viewer, creates the viewer if not provided, and returns any added layers)
2. modify the return of each of the `view_*` functions to return the first item in the tuple (the viewer) to keep their APIs consistent:
    ```python
        return _make_viewer_then('add_labels', args, kwargs)[0]
    ```

3. Now imshow just becomes:
```python
def imshow(*args, viewer=None, **kwargs):
    return _make_viewer_then('add_image', args, kwargs, viewer=viewer)
```

File: napari/view_layers.py
Comment (liu-ziyang) [inline napari/view_layers.py]: need to update the docstring to reflect the new behavior

**Comment by kcpevey** [comment]
Should I also adapt `_merge_layer_viewer_sigs_docs` to work for `imshow` as well?

**Comment by tlambert03** [comment]
> Should I also adapt _merge_layer_viewer_sigs_docs to work for imshow as well?

thanks @kcpevey!  that `_merge_layer_viewer_sigs_docs` function was actually the one I was mostly thinking of when I said "let's avoid the autogeneration".  Dynamically generating signatures is the ickiest bit of what we're doing in this module, and if we're ultimately thinking of moving away from the `view_*`  functions (I think?) then maybe we should just manually write the signature and docs of `imshow()`?  It does returning to duplication between the docs in `Viewer` and `Image` ... but it's a bit more palatable if it's just a single imshow function rather than 7 or 8 `view` functions.

So, I think I'm proposing that you go ahead and do the laborious thing and manually write out all of the arguments in the `imshow` signature and doc strings, but then just internally pass them all to `_make_viewer_then` as you've done here.

That way, it will be very easy to keep the signature you've created for `imshow` when we later pull `channel_axis` out of `add_image()` (and just rework the internals if/when we need to)

what do @sofroniewn and @jni think?  (i could go either way here)

**Comment by liu-ziyang** [comment]
(I am not asked to be here, but I second that we should remove autogeneration, having these entry points functions being autogenerated makes debugging stack tracing more difficult. Plus the code generated is not that much, if we are talking about a few thousands of lines, sure. But this is just a handful of methods that we are saving up)

**Review by liu-ziyang — APPROVED** [review]
I wish we go a bit further in removing the autogeneration, but not going to block this PR. The changes here to address the issue looks fine to my eyes for now.

**Comment by tlambert03** [comment]
> I wish we go a bit further in removing the autogeneration, but not going to block this PR. 

yeah, that's definitely for another PR, since we need to essentially deprecate and then remove all of the `view_*` methods first!

File: napari/view_layers.py
Comment (tlambert03) [inline napari/view_layers.py:254]: `channel_axis` is in the tests, but isn't mentioned in the kwargs here.  Since this new function [has been proposed](https://github.com/napari/napari/pull/4858#issuecomment-1200505333) as the thing we'll use to deprecate the `channel_axis` argument in `add_image`, we definitely need that represented here.

File: napari/view_layers.py
Comment (tlambert03) [inline napari/view_layers.py]: I'm afraid kwargs here should also be laboriously listed out... as we've done for `add_image` ... and it should also include the `viewer.__init__ ` stuff.  Essentially, it needs to manually do what `_merge_layer_viewer_sigs_docs` was doing (which was to build _both_ the signature and the docs.

```python
    def imshow(
        data,
        *,
        channel_axis=None,
        rgb=None,
        colormap=None,
        contrast_limits=None,
        gamma=1,
        interpolation2d='nearest',
        interpolation3d='linear',
        ...
```

using **kwargs one loses all the tab-autocompletion in an interactive terminal or in an IDE.  this is course is super annoying, and it also leads to the possibility that these defaults go out of sync with `add_image`... (and this is why we've autogenerated these in the past).  so, we might also want to add back the test to make sure that all the parameters match.  You can see an old test we used to use here:  https://github.com/napari/napari/blob/871825e9ca9f570575e0681132d638e03cb39708/napari/_tests/test_viewer_layer_parity.py

While I do think the signature should be updated here, I'll let others weigh in on whether to bring back those old tests (which were always a bit annoying TBH).  

Also, @jni should be signing off on this PR as well since it was his proposal

File: napari/_tests/test_view_layers.py
Comment (jni) [inline napari/_tests/test_view_layers.py]: I think we should test that the returned viewer is the same that was passed in, and that the layer data is the same as as the data that was given:

```suggestion
    viewer = make_napari_viewer()
    viewer2, layer = napari.imshow(data, viewer=viewer, show=False)
    assert viewer is viewer2
    np.testing.assert_array_equal(data, layer.data)
```

I use np testing in the suggestion in case we ever decide to wrap the NumPy arrays e.g. with an evented NumPy array.

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]: Since we're at it, this needs a full docstring I think... The title and type annotations are not enough for me to understand what it's doing —

```suggestion
def _make_viewer_then(
    add_method: str, args, kwargs, viewer=None
) -> Tuple[Viewer, Any]:
    """Create a viewer, call given add_* method, then return viewer and layer.

    Parameters
    ----------
    add_method : str
        Which ``add_`` method to call on the viewer, e.g. `add_image`,
        or `add_labels`.
    args : list
        Positional arguments for the ``add_`` method.
    kwargs : dict
        Keyword arguments for either the `Viewer` constructor or for the
        ``add_`` method.
    viewer : Viewer, optional
        A pre-existing viewer, which will be used provided, rather than
        creating a new one.
        
    Returns
    -------
    viewer : napari.Viewer
        The created viewer, or the same one that was passed in, if given.
    layer(s): napari.layers.Layer or List[napari.layers.Layer]
        The value returned by the add_method. Can be a list of layers if
        ``add_image`` is called with a ``channel_axis=`` keyword
        argument.
    """
```

File: napari/view_layers.py
Comment (jni) [reply napari/view_layers.py]: I'm +1 on the full copy (yes, annoying) and on the tests, too. I think they'll be quite critical tbh as we start to whittle down the signatures of all the add_ methods.

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]: Assuming we want to keep the channel_axis= behaviour (I think we decided yes), then I think maybe we want this to *always* return a list of layers, which would contain a single layer if `channel_axis=None`, or one or more if `channel_axis=some_int`. Ref:

https://github.com/napari/napari/issues/4693#issuecomment-1200509018

> where layers is a list (not a LayerList) of any added Layer objects - will be 1 if multichannel=False and will be more if multichannel=True.

Thoughts @tlambert03? ie do we want the return type to be variable, or always a list (though it may be cumbersome in the single channel case)? e.g. is this too clunky?

```
viewer, (layer,) = napari.imshow(array)
```

Maybe it is and we leave it as variable?

**Review by jni — COMMENTED** [review]
Thanks for taking this on @kcpevey! Very cool to come back from holidays and see one of your issues getting fixed! 😃 I've left some minor comments and one major decision to be made... Probably something for the core team to discuss before deciding...

File: napari/view_layers.py
Comment (jni) [inline napari/view_layers.py]:
```suggestion
    viewer : napari.Viewer
        The created or passed viewer.
    layer(s) : napari.layers.Image or List[napari.layers.Image]
        The added layer(s). (May be more than one if the ``channel_axis`` keyword
        argument is given.
```

File: napari/view_layers.py
Comment (tlambert03) [reply napari/view_layers.py]: > Thoughts @tlambert03? ie do we want the return type to be variable, or always a list (though it may be cumbersome in the single channel case)? e.g. is this too clunky?

my vote would definitely be to always return a list (i think it's well worth the minor clunkiness) ... and I'd like to see a type annotation on the return as well ` -> Tuple[Viewer, List[Image]]`.  

but I think that check should be done here, rather than in `_make_viewer_then`. 

```python
viewer, layers = _make_viewer_then('add_image', args, kwargs, viewer=viewer)
layers = layers if isinstance(layers, list) else [layers]
return viewer, layers
```

File: napari/view_layers.py
Comment (tlambert03) [reply napari/view_layers.py]: no harm in adding this here, but it could also contain a note that this is the function we might be killing as we get rid of `view_*`...   And when the last `view_*` function is removed, then the body of this function would just move below to `imshow` (which would be the only remaining function to use this logic)

File: napari/view_layers.py
Comment (kcpevey) [reply napari/view_layers.py:212]: Isn't `channel_axis` only used by `add_image`? If so, I guess its fine to include if I mention that its for image only?

Comment (kcpevey) [reply napari/view_layers.py:254]: `channel_axis` is not in the `ImageBase` signature. Its only added [here](https://github.com/napari/napari/blob/main/napari/components/viewer_model.py#L598). Why is that? What is the purpose of the method in the viewer_model?

Comment (kcpevey) [reply napari/view_layers.py:254]: also, a couple of the default values in viewer_model are different from the ImageBase default values

File: napari/view_layers.py
Comment (jni) [reply napari/view_layers.py]: It's funny, overnight I was thinking, nah, let's keep it variable, the whole point of a convenience function is to be convenient, not to be type-safe... 😂 Especially given the motivation for this whole thing from #4693 (existing convenience functions are not quite convenient enough)... How strongly do you feel about it @tlambert03?

Comment (tlambert03) [reply napari/view_layers.py]: Eh, 6 out of 10 strong?  

I think it's convenient up until the point where you get a type error because you didn't switch over your return object handling when you changed your channel axis argument.  And then it would have been more convenient not to have to remember to change the rest of your code accordingly.

Yes, a big driver for this is convenience, but it kinda got pushed back to the fore because of the channel axis deprecation, which is largely because of the type problem.

But it's not a blocker for me, so you can make the call :)

Comment (jni) [reply napari/view_layers.py]: Ok I'm back 😂 

The type thingy is interesting, because I would probably write a lot of `image, (layer,) = ...`, and then you're in the same box as before. (Though in this case the error happens immediately, rather than when trying to access layer attributes later... That's an advantage.) I would argue that in code (vs interactive usage), we would discourage the use of imshow in favour of instantiating a viewer and using `add_` methods.

I'd love to hear from other core devs and users on this... I'll ping on Zulip.

Comment (tlambert03) [reply napari/view_layers.py]: As I said, not a blocker for me. So if you're coming around to a strong opinion for variable return types, that's fine

Comment (jni) [reply napari/view_layers.py]: After a hearty discussion on Zulip we have two options, either variable return type or variable *length* return type, `return viewer, *layers`. [Here's the summary message on Zulip](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/imshow.20return.20type/near/296525428). People are going to vote on the options and then we'll settle on one. I'm happy with either.

**Comment by alisterburt** [comment]
Following [discussion on zulip](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/imshow.20return.20type/near/296525428) it seems the 'multiple dispatch' option has won out! @kcpevey it would be great to get this in over the next week so that it can be in the hands of users for 0.4.17. 🥳  

Please ping me for review when you feel this is ready for another look before going in!

**Review by jni — APPROVED** [review]
@alisterburt I think this is ready to go, actually! It turns out the multiple dispatch option is what @kcpevey had implemented originally anyway. 😊 And as far as I can tell, all other reviewer comments have been addressed.

File: napari/view_layers.py
Comment (brisvag) [reply napari/view_layers.py]: @jni, @alisterburt this was not yet addressed. Should we tackle it before merging?

**Review by brisvag — DISMISSED** [review]

File: napari/view_layers.py
Comment (alisterburt) [reply napari/view_layers.py]: Yes absolutely - I'm afk until at least tomorrow but will do then if not done before!

**Comment by kcpevey** [comment]
I was waiting for the return type discussion to make some final changes. This is now ready for final review (pending one question). @jni @brisvag @alisterburt

File: napari/view_layers.py
Comment (kcpevey) [inline napari/view_layers.py]: I'm not sure how to handle this. Can I just remove it? Or can someone make a suggestion?

Comment (brisvag) [reply napari/view_layers.py]: As far as I understand, the intention was to unify this with the other layers (i.e: no channel splitting stuff). However, I don't think we wanted to release such a big break in VERY public api in a minor release. Correct me if I'm wrong @jni?

If I'm right, then I would say deprecate the use of `channel_axis` (and any other related kwarg) in add_image, and then remove this whole overloaded function in 0.5.0.

File: napari/view_layers.py
Comment (alisterburt) [inline napari/view_layers.py]:
```suggestion
```

**Comment by alisterburt** [comment]
@kcpevey based on what @brisvag says I think we can safely remove the commented line so have done - let me know if that's not the case

@brisvag I think your requested changes have been addressed, could you unblock here please?

**Comment by brisvag** [comment]
> @brisvag I think your requested changes have been addressed, could you unblock here please?

I think we still need to deprecate the use of `channel_axis` in `add_image` and raise a warning. This will be a big change in `0.5.0`, so it's good to warn now if we can!

**Comment by kcpevey** [comment]
> As far as I understand, the intention was to unify this with the other layers (i.e: no channel splitting stuff).

> I think we still need to deprecate the use of channel_axis in add_image and raise a warning.

I agree if channel_axis is going to be deprecated, we should raise a warning. However, I don't see how the layers will be unified into a single function. I can't see how channel_axis _could_ be deprecated in add_image without losing functionality. I just don't know the codebase well enough. I will let you guys drive the decision on this.

**Comment by brisvag** [comment]
> I can't see how channel_axis _could_ be deprecated in add_image without losing functionality

I might be conflating a bit with the discussion in #4858. In my mind, these are tightly linked (add imshow and remove special case of image).

**Comment by jni** [comment]
@brisvag the motivation for imshow is *partly* linked to the deprecation of `channel_axis`. But imshow is useful on its own and can happen independently. So I'm going to merge this and leave the deprecation to its own PR.

@kcpevey 

> I can't see how channel_axis could be deprecated in add_image without losing functionality

Indeed, we would lose the functionality from `add_image`, *but* it could be regained with a for-loop or by using your new imshow function. The key is that we want *all* `viewer.add_*` methods to always have the same effect: add one layer, and return it. `channel_axis` breaks that.

**Comment by jni** [comment]
Thank you @kcpevey for implementing this! ❤️

**Pull request merged**
